src/nodes.coffee

```diff
@@ -2247,8 +2250,16 @@ exports.Assign = class Assign extends Base
 
       restElements
 
-    # Cache the value for reuse with rest elements
-    [@value, valueRef] = @value.cache o
+    # Cache the value for reuse with rest elements.
+    # `Obj` should be always cached.
+    # Examples:
+    #   {a, r...} = {a:1, b:2, c:3}
+    #   {a, r...} = {a:1, obj...}
+    shouldCache = (value) ->
+      return yes if value.base instanceof Obj
```

Comment:
I've already tried that, but then tests fail.
I also think object destructuring with spreads is the only place where this is necessary.
Besides, spreads might soon reach stage-4.